cocotb/monitors/avalon.py

```diff
@@ -100,15 +100,20 @@ class AvalonSTPkts(BusMonitor):
         "invalidTimeout"                : 0,
     }
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, entity, name, clock, **kwargs):
         config = kwargs.pop('config', {})
-        BusMonitor.__init__(self, *args, **kwargs)
+        report_channel = kwargs.pop('report_channel', False)
+        BusMonitor.__init__(self, entity, name , clock, **kwargs)
 
         self.config = self._default_config.copy()
+        self.report_channel = report_channel
 
         # Set default config maxChannel to max value on channel bus
         if hasattr(self.bus, 'channel'):
             self.config['maxChannel'] = (2 ** len(self.bus.channel)) -1
+        else:
+            if report_channel:
+                self.log.error("Channel reporting asked on bus without channel signal")
```

Comment:
This should be `raise ValueError(...)`

Reply:
OK, that was original version of patch but changed it to use self.log.error. Any guidance on when the raise and when to use self.log.error ?

Reply:
Use `raise` if the caller called your function in an invalid way, and it doesn't make sense to continue.

Use `self.log.error` if the bus itself is misbehaving, and throwing an error immediately would leave the bus in an invalid state. Even if you do eventually throw an exception, if you weren't able to do it immediately, you should also `log.error` so that the sim time of when things went wrong is recorded.

TL;DR: `log.error` is for humans only, `raise` is for both humans and code.

Reply:
> TL;DR: `log.error` is for humans only, `raise` is for both humans and code.

This text should be part of a "How to write cocotb code", maybe in #1141.